### PR TITLE
`azurerm_shared_image_gallery` - support for the sharing block

### PR DESCRIPTION
**Pull request by myc2h6o**
To add support for Community Shared Gallery.
Added tests as well for VM to consume Community Shared Image Version:
https://github.com/hashicorp/terraform-provider-azurerm/blob/f70fe79ef6f7464b191c9cbb3aa46ab064089560/internal/services/compute/linux_virtual_machine_resource.go#L302
There will be another possible value for `sharing_profile.permission` to support Direct Shared Gallery after its GA so  `community_gallery_info` is set to `optional`

### CODE REVIEW

**Comment by myc2h6o** [comment]
Test result:
image

File: internal/services/compute/shared_image_gallery_resource.go [hunk 1]
```diff
@@ -53,6 +55,64 @@ func resourceSharedImageGallery() *pluginsdk.Resource {
 				Optional: true,
 			},
 
+			"sharing_profile": {
```
Comment (WodansSon) [inline internal/services/compute/shared_image_gallery_resource.go:58]:
```suggestion
			"sharing": {
```
Per the [reference naming](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/reference-naming.md) doc should we shorten this to just `sharing`?

File: internal/services/compute/shared_image_gallery_resource.go [hunk 2]
```diff
@@ -53,6 +55,64 @@ func resourceSharedImageGallery() *pluginsdk.Resource {
...
+						"permission": {
+							Type:     pluginsdk.TypeString,
+							Required: true,
+							ForceNew: true,
+							ValidateFunc: validation.StringInSlice([]string{
+								string(galleries.GallerySharingPermissionTypesCommunity),
+							}, false),
+						},
+
+						"community_gallery_info": {
```
Comment (WodansSon) [inline internal/services/compute/shared_image_gallery_resource.go:74]:
```suggestion
						"community_gallery": {
```
Same here, as mentioned above, should we shorten this to just `community_gallery`?

File: internal/services/compute/shared_image_gallery_resource.go [hunk 3]
```diff
@@ -53,6 +55,64 @@ func resourceSharedImageGallery() *pluginsdk.Resource {
...
+										ForceNew:     true,
+										ValidateFunc: validation.StringIsNotEmpty,
+									},
+									"publisher_uri": {
+										Type:         pluginsdk.TypeString,
+										Required:     true,
+										ForceNew:     true,
+										ValidateFunc: validation.StringIsNotEmpty,
+									},
+									"public_name": {
```
Comment (WodansSon) [inline internal/services/compute/shared_image_gallery_resource.go:105]:
```suggestion
									"name": {
```
Should this be shortened to just `name`?

File: internal/services/compute/shared_image_gallery_resource.go [hunk 4]
```diff
@@ -53,6 +55,64 @@ func resourceSharedImageGallery() *pluginsdk.Resource {
...
+							MaxItems: 1,
+							Elem: &pluginsdk.Resource{
+								Schema: map[string]*pluginsdk.Schema{
+									"eula": {
+										Type:         pluginsdk.TypeString,
+										Required:     true,
+										ForceNew:     true,
+										ValidateFunc: validation.StringIsNotEmpty,
+									},
+									"public_name_prefix": {
```
Comment (WodansSon) [inline internal/services/compute/shared_image_gallery_resource.go:87]:
```suggestion
									"prefix": {
```
Would it make more sense if this was just `prefix` or `name_prefix`?

File: internal/services/compute/shared_image_gallery_resource.go [hunk 5]
```diff
@@ -65,6 +125,7 @@ func resourceSharedImageGallery() *pluginsdk.Resource {
 
 func resourceSharedImageGalleryCreateUpdate(d *pluginsdk.ResourceData, meta interface{}) error {
```
Comment (WodansSon) [inline internal/services/compute/shared_image_gallery_resource.go:126]: Since update is now supported I believe we should make this two separate functions (e.g. `resourceSharedImageGalleryCreate` and `resourceSharedImageGalleryUpdate`)

File: internal/services/compute/shared_image_gallery_resource.go [hunk 6]
```diff
@@ -154,9 +232,120 @@ func resourceSharedImageGalleryDelete(d *pluginsdk.ResourceData, meta interface{
...
 	}
 
+	resp, err := client.Get(ctx, *id, galleries.DefaultGetOperationOptions())
+	if err != nil {
+		return fmt.Errorf("retrieving %s: %+v", id, err)
+	}
+
+	if model := resp.Model; model != nil {
+		if prop := model.Properties; prop != nil && prop.SharingProfile != nil && prop.SharingProfile.Permissions != nil {
+			if *prop.SharingProfile.Permissions == galleries.GallerySharingPermissionTypesCommunity {
```
Comment (WodansSon) [inline internal/services/compute/shared_image_gallery_resource.go:242]:
```suggestion
			if pointer.From(prop.SharingProfile.Permissions) == galleries.GallerySharingPermissionTypesCommunity {
```
To be consistent in the code should we migrate this to use the `pointer` helper? 

NOTE: I am not going to mark all of the `*` code, but there are a lot of places in the code where this can be migrated as well. WDYT?

File: website/docs/r/shared_image_gallery.html.markdown [hunk 7]
```diff
@@ -44,8 +44,32 @@ The following arguments are supported:
...
 
+---
+
+A `sharing_profile` block supports the following:
+
+* `permission` - (Required) The permission of the Shared Image Gallery when sharing. The only possible value now is `Community`. Changing this forces a new resource to be created.
+
+* `community_gallery_info` - (Optional) A `community_gallery_info` block as defined below. Changing this forces a new resource to be created.
+
+~> **NOTE** `community_gallery_info` must be set when `permission` is set to `Community`.
```
Comment (WodansSon) [inline website/docs/r/shared_image_gallery.html.markdown:59]:
```suggestion
~> **NOTE:** `community_gallery_info` must be set when `permission` is set to `Community`.
```
Super minor, but should we add a colon `:` at the end of `NOTE`?

**Review by WodansSon — CHANGES_REQUESTED** [review]
@myc2h6o thank you so much for this PR. This is looking really good so far, but I did leave a few comments. If you can take a look at those I will be happy to give this a second review to get this merged! 🚀

**Comment by myc2h6o** [comment]
Hi @WodansSon thanks for reviewing the change! I've updated the pr according to your comment, please take a look.

**Comment by myc2h6o** [comment]
Updated test result:
image

File: website/docs/r/shared_image_gallery.html.markdown [hunk 8]
```diff
@@ -44,8 +44,32 @@ The following arguments are supported:
...
 
+* `sharing` - (Optional) A `sharing` block as defined below. Changing this forces a new resource to be created.
+
 * `tags` - (Optional) A mapping of tags to assign to the Shared Image Gallery.
 
+---
+
+A `sharing` block supports the following:
+
+* `permission` - (Required) The permission of the Shared Image Gallery when sharing. The only possible value now is `Community`. Changing this forces a new resource to be created.
```
[the remark this answers is not in the document]
Comment (myc2h6o) [reply website/docs/r/shared_image_gallery.html.markdown:55]: thanks for pointing that out, I've updated the document. This shall be GAed already in most of the regions, including the regions used by our acc tests so test subscriptions shall be fine to run the new test
